internal/batches/executor/run_steps.go

```diff
@@ -25,47 +27,61 @@ import (
 	yamlv3 "gopkg.in/yaml.v3"
 )
 
-type runStepsOpts struct {
-	// wc is the workspace creator to use. It will be called at the beginning of
+type RunStepsOpts struct {
+	// WC is the workspace creator to use. It will be called at the beginning of
 	// runSteps to prepare the workspace for execution.
-	wc workspace.Creator
-	// ensureImage is called in runSteps to make sure the used image has been
+	WC workspace.Creator
+	// EnsureImage is called in runSteps to make sure the used image has been
 	// pulled from the registry.
-	ensureImage imageEnsurer
-	// task is the definition of the workspace execution.
-	task *Task
-	// tempDir points to where temporary files of the execution should live at.
-	tempDir string
-
-	logger log.TaskLogger
-
-	ui StepsExecutionUI
-
-	// isRemote toggles server-side execution mode. This disables file mounts using
+	EnsureImage imageEnsurer
+	// Task is the definition of the workspace execution.
+	Task *Task
+	// TempDir points to where temporary files of the execution should live at.
+	TempDir string
+	// Timeout sets the deadline for the execution context. When exceeded,
+	// execution will stop and an error is returned.
+	Timeout time.Duration
+	// RepoArchive is the repo archive to be used for creating the workspace.
+	RepoArchive repozip.Archive
+	Logger      log.TaskLogger
+	UI          StepsExecutionUI
+	// IsRemote toggles server-side execution mode. This disables file mounts using
 	// step.mounts.
-	isRemote bool
-
-	// globalEnv is the os.Environ() for the execution. We don't read from os.Environ()
+	IsRemote bool
+	// GlobalEnv is the os.Environ() for the execution. We don't read from os.Environ()
 	// directly to allow injecting variables and hiding others.
-	globalEnv []string
+	GlobalEnv []string
 }
 
-func runSteps(ctx context.Context, opts *runStepsOpts) (stepResults []execution.AfterStepResult, err error) {
-	opts.ui.ArchiveDownloadStarted()
-	err = opts.task.RepoArchive.Ensure(ctx)
-	opts.ui.ArchiveDownloadFinished(err)
+func RunSteps(ctx context.Context, opts *RunStepsOpts) (stepResults []execution.AfterStepResult, err error) {
+	// Set up our timeout.
```

Comment:
Moved to here from the executor so timeouts work when using runsteps directly, too.